linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/StringPool.scala

```diff
@@ -20,85 +20,75 @@ import org.scalajs.linker.backend.webassembly.Instructions._
   }
 
-  /** Returns the list of instructions that load the given constant string.
-   *
-   *  The resulting list is *not* a Wasm constant expression, since it includes
-   *  a `call` to the helper function `stringLiteral`.
+  /** Generates the string pool, and returns the list of WTF-16 strings that
+   *  must be provided by the JS embedding.
    */
-  def getConstantStringInstr(str: String): List[Instr] =
-    getConstantStringDataInstr(str) :+ Call(genFunctionID.stringLiteral)
-
-  /** Returns the list of 3 constant integers that must be passed to `stringLiteral`.
-   *
-   *  The resulting list is a Wasm constant expression, and hence can be used
-   *  in the initializer of globals.
-   */
-  def getConstantStringDataInstr(str: String): List[I32Const] = {
-    val data = register(str)
-    List(
-      I32Const(data.offset),
-      I32Const(str.length()),
-      I32Const(data.constantStringIndex)
-    )
-  }
-
-  def genPool()(implicit ctx: WasmContext): Unit = {
+  def genPool()(implicit ctx: WasmContext): List[(String, String)] = {
     poolWasGenerated = true
 
-    ctx.moduleBuilder.addData(
-      Data(
-        genDataID.string,
-        OriginalName("stringPool"),
-        rawData.toArray,
-        Data.Mode.Passive
-      )
-    )
-
-    ctx.addGlobal(
-      Global(
-        genGlobalID.stringLiteralCache,
-        OriginalName("stringLiteralCache"),
-        isMutable = false,
-        RefType(genTypeID.externrefArray),
-        Expr(
-          List(
-            I32Const(nextIndex), // number of entries in the pool
-            ArrayNewDefault(genTypeID.externrefArray)
+    val wtf16Strings = new mutable.ListBuffer[(String, String)]
+
+    for (str <- registeredStrings.toSeq.sorted) {
+      if (isValidUTF16String(str)) {
+        // Magic import from the JS string builtins proposal
+        ctx.moduleBuilder.addImport(
+          Import(
+            UTF8StringConstantsModule,
+            str, // by definition, the imported member name is the string itself
+            ImportDesc.Global(genGlobalID.forStringLiteral(str),
+                OriginalName("'" + str), isMutable = false, RefType.extern)
           )
         )
-      )
-    )
+      } else {
+        // WTF-16 strings, which we import from a generated JS object
+        val indexStr = wtf16Strings.size.toString()
+        wtf16Strings += indexStr -> str
+        ctx.moduleBuilder.addImport(
+          Import(
+            WTF16StringConstantsModule,
+            indexStr,
+            ImportDesc.Global(genGlobalID.forStringLiteral(str),
+                OriginalName("wtf16." + indexStr), isMutable = false, RefType.extern)
+          )
+        )
+      }
+    }
+
+    wtf16Strings.toList
+  }
+
+  private def isValidUTF16String(str: String): Boolean = {
+    // scalastyle:off return
+    val len = str.length()
+    var i = 0
+    while (i != len) {
+      val cp = str.codePointAt(i)
+      if (cp >= Character.MIN_SURROGATE && cp <= Character.MAX_SURROGATE)
```

Comment:
```suggestion
      if (Character.isSurrogate(cp))
```

For readability?

Reply:
`Character.isSurrogate` only accepts a `Char` :( I added a comment.